src/devtools/client/inspector/rules/components/RulesApp.tsx

```diff
@@ -63,120 +70,123 @@ export const RulesApp: FC<RulesAppProps> = ({
-            {inheritedSource}
-          </div>
+          <Rule
+            key={`${inheritedNodeId}|${rule.id}`}
+            rule={rule}
+            {...ruleProps}
+            query={rulesQuery}
+          />
         );
       }
 
-      output.push(
-        <Rule key={`${inheritedNodeId}|${rule.id}`} rule={rule} {...ruleProps} query={rulesQuery} />
-      );
-    }
-
-    return output;
-  };
-
-  const renderStyleRules = (rules: RuleState[]) => {
-    if (!rules.length) {
-      return null;
-    }
-
-    return <Rules {...ruleProps} rules={rules} />;
-  };
+      return output;
+    },
+    [rulesQuery, ruleProps]
+  );
 
-  const renderPseudoElementRules = (rules: RuleState[]) => {
-    if (!rules.length) {
-      return null;
-    }
+  const renderStyleRules = useCallback(
+    (rules: RuleState[]) => {
+      return <Rules {...ruleProps} query={rulesQuery} rules={rules} />;
+    },
+    [rulesQuery, ruleProps]
+  );
 
-    type FCProps<C> = C extends FC<infer P> ? P : never;
-    const componentProps: FCProps<typeof Rules> = {
-      rules,
-      ...ruleProps,
-    };
-
-    const items = [
-      {
-        component: Rules,
-        componentProps,
-        header: getStr("rule.pseudoElement"),
-        id: "rules-section-pseudoelement",
-        opened: Services.prefs.getBoolPref(SHOW_PSEUDO_ELEMENTS_PREF),
-        onToggle: (opened: boolean) => {
-          Services.prefs.setBoolPref(SHOW_PSEUDO_ELEMENTS_PREF, opened);
+  const renderPseudoElementRules = useCallback(
+    (rules: RuleState[]) => {
+      type FCProps<C> = C extends FC<infer P> ? P : never;
+      const componentProps: FCProps<typeof Rules> = {
+        rules,
+        query: rulesQuery,
+        ...ruleProps,
+      };
+
+      const items = [
+        {
+          component: Rules,
+          componentProps,
+          header: getStr("rule.pseudoElement"),
+          id: "rules-section-pseudoelement",
+          opened: Services.prefs.getBoolPref(SHOW_PSEUDO_ELEMENTS_PREF),
+          onToggle: (opened: boolean) => {
+            Services.prefs.setBoolPref(SHOW_PSEUDO_ELEMENTS_PREF, opened);
+          },
         },
-      },
-    ];
+      ];
 
-    return (
-      <>
-        <Accordion items={items} />
-        <div className="ruleview-header">{getStr("rule.selectedElement")}</div>
-      </>
-    );
-  };
+      return (
+        <>
+          <Accordion items={items} />
+          <div className="ruleview-header">{getStr("rule.selectedElement")}</div>
+        </>
+      );
+    },
+    [ruleProps, rulesQuery]
+  );
 
   const rulesElements = useMemo(() => {
     if (!rulesQuery && rules.length === 0) {
       return <div className="devtools-sidepanel-no-result">{getStr("rule.empty")}</div>;
     }
 
-    const inheritedRules = [];
-    const pseudoElementRules = [];
-    const styleRules = [];
+    const nonEmptyRules = rules.filter(rule =>
+      rule.declarations.some(declaration => !declaration.isInvisible)
+    );
```

Comment:
The rest of the changes are more on the side of "maintenance", but this is the actual change that saves a huge number of rules from being unnecessarily rendered. This change seems to be on par with what browsers render in their own DevTools panels, so this is not "a hack".

Reply:
Smart